Ibovespa index support

igor17400 opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
- For the **feat: download ibovespa index historic composition**  the following files were included:
  - `scripts/data_collector/br_index/README.md`
  - `scripts/data_collector/br_index/collector.py`
  - `scripts/data_collector/br_index/requirements.txt`
- For **fix: typo error instead of end_date, it was written end_ate** the following files were modified:
  - `scripts/data_collector/index.py`
- For **feat: adds support for downloading stocks historic prices from Brazil's stocks exchange (B3)** the following files were modified:
  - `scripts/data_collector/utils.py`
  - `scripts/data_collector/yahoo/README.md`
  - `scripts/data_collector/yahoo/collector.py`


## Motivation and Context

Brazil's stock exchange is one of the biggest in the world, therefore adding its dataset into qlib can help people do research via qlib infrastructure using different datasets.

1. Application scenario: support brazilian stocks exchange investment strategy research.
2. Related works (Papers, Github repos etc.): Not applicable
3. Any other relevant and important information: More details on Brazil's stock exchange can be found on [B3](https://www.b3.com.br/en_us/)

## How Has This Been Tested?
- [x] Pass the test by running: `pytest qlib/tests/test_all_pipeline.py` under upper directory of `qlib`.

- [x] If you are adding a new feature, test on your own test scripts. 

There isn't a specific script for test, however the added code can be tested running the following commands:

- `python scripts/data_collector/br_index/collector.py --index_name IBOV --qlib_dir ~/.qlib/qlib_data/br_data --method parse_instruments`
- `python scripts/data_collector/br_index/collector.py --index_name IBOV --qlib_dir ~/.qlib/qlib_data/br_data --method save_new_companies`
- `python scripts/data_collector/br_index/collector.py download_data --source_dir ~/.qlib/stock_data/source/br_data --start 2003-01-03 --end 2022-03-01 --delay 1 --interval 1d --region BR`
- `python scripts/data_collector/br_index/collector.py download_data --source_dir ~/.qlib/stock_data/source/br_data_1min --delay 1 --interval 1min --region BR`

<!--- **ATTENTION**: If you are adding a new feature, please make sure your codes are **correctly tested**. If our test scripts do not cover your cases, please provide your own test scripts under the `tests` folder and test them. More information about test scripts can be found [here](https://docs.python.org/3/library/unittest.html#basic-example), or you could refer to those we provide under the `tests` folder. -->

## Screenshots of Test Results (if appropriate):
1. Pipeline test: Not applicable
2. Your own tests: Not applicable

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Add new feature
- [x] Update documentation (README.md)

closes #956

ChiahungTai commented:
@you-n-g Should it rerun the mac os test?

you-n-g commented on scripts/data_collector/br_index/collector.py:
```diff
@@ -0,0 +1,317 @@
+    "2021_1Q",
+    "2021_2Q",
+    "2021_3Q",
+    "2022_1Q",
```
Hi,
Why don't use a for loop to generate this list  until  `datetime.datetime.today()`

igor17400 replied:
Great!
Thanks for the suggestion
I'll work on this change

you-n-g commented on scripts/data_collector/br_index/collector.py:
```diff
@@ -0,0 +1,317 @@
+
+IBOV_INDEX_COMPOSITION = "https://raw.githubusercontent.com/igor17400/IBOV-HCI/main/historic_composition/{}.csv"
+
+YEAR_QUARTER = [
```
I think this will be more reasonable to make it as an class attribute  of IBOVIndex

igor17400 replied:
Ook

you-n-g commented on scripts/data_collector/utils.py:
```diff
@@ -19,6 +19,7 @@
 from tqdm import tqdm
 from functools import partial
 from concurrent.futures import ProcessPoolExecutor
+from bs4 import BeautifulSoup
```
Maybe BeautifulSoup should be in the `requirements.txt` ?

SunsetWolf commented:
In the course of reviewing this PR we found two issues.
1. there are multiple places where the get_instruments() method is used, and we feel that scripts.index.py is the best place for the get_instruments() method to go.
2. data_collector.utils has some very generic stuff put inside it.
Are you willing to fix them? If not, I will fix these issues in the future.

igor17400 commented:
@SunsetWolf I can work on this!
I'll give it a look

igor17400 commented:
I was checking some functions in the Yahoo `collector.py` code and I suspect I may have found another issue.


> The callable method returns True if the object passed appears callable. If not, it returns False. As the image below shows,

![Screen Shot 2022-03-31 at 19 05 43](https://user-images.githubusercontent.com/42355184/161156939-e98717be-4ef0-4f57-b36d-bb97ddcb6d38.png)

I was trying to set the `retry` parameter from the decorator `deco_retry` to 1. Because, Yahoo finance does not keeps track of the majority of Brazilian stocks and as a consequence the code is taking a long time to finish. And I was doing so by calling the decorator as `@deco_retry(retry=1)` and defining `retry` variable inside the `_get_ibovespa()` function as the code snipet below shows.

However, it seems that the code below for the decorator `deco_retry` will never return `True` for `callable(retry)` since `retry` variable is an `int` which is not callable by its own nature.

**_Am I doing something wrong?_**

```
def deco_retry(retry: int = 5, retry_sleep: int = 3):
    def deco_func(func):
        @functools.wraps(func)
        def wrapper(*args, **kwargs):
            _retry = 5 if callable(retry) else retry
            _result = None
            for _i in range(1, _retry + 1):
                try:
                    _result = func(*args, **kwargs)
                    break

                except Exception as e:
                    logger.warning(f"{func.__name__}: {_i} :{e}")
                    if _i == _retry:
                        raise

                time.sleep(retry_sleep)
            return _result

        return wrapper

    return deco_func(retry) if callable(retry) else deco_func
```

```
 @deco_retry(retry=1)
    def _get_ibovespa():
        retry = 1
        _symbols = []
        url = "https://www.fundamentus.com.br/detalhes.php?papel="

        # Request
        agent = {"User-Agent": "Mozilla/5.0"}
        page = requests.get(url, headers=agent)

        # BeautifulSoup
        soup = BeautifulSoup(page.content, "html.parser")
        tbody = soup.find("tbody")

        children = tbody.findChildren("a", recursive=True)
        for child in children:
            _symbols.append(str(child).split('"')[-1].split(">")[1].split("<")[0])

        return _symbols
```

you-n-g commented on scripts/data_collector/us_index/collector.py:
```diff
@@ -269,46 +270,6 @@ def parse_instruments(self):
-
 
 if __name__ == "__main__":
     fire.Fire(get_instruments)
```
```suggestion
    fire.Fire(partial(get_instruments, market_index="br_index" ))
```
- This applies to other index data source as well.


Do you think this will make the interface easier?
Then all the  collector.py CLI  in each  folder can remove a redundant arguments.

you-n-g commented on scripts/data_collector/br_index/README.md:
````diff
@@ -53,9 +53,9 @@ With that reference, the index's composition can be compared quarter by quarter
 ```bash
 # parse instruments, using in qlib/instruments.
-python collector.py --index_name IBOV --qlib_dir ~/.qlib/qlib_data/br_data --method parse_instruments
+python collector.py --index_name IBOV --qlib_dir ~/.qlib/qlib_data/br_data --method parse_instruments --market_index br_index
````
![image](https://user-images.githubusercontent.com/465606/161372029-c9410b46-1a65-4442-8a06-ce4b9a1cbc05.png)

you-n-g commented on scripts/data_collector/utils.py:
```diff
@@ -564,3 +562,47 @@ def generate_minutes_calendar_from_daily(
 
 if __name__ == "__main__":
```
It will be better if `ifmain` is at the bottom of the script.

igor17400 replied:
Yes, much better!
It was a lack of attention on my part

you-n-g commented on scripts/data_collector/yahoo/collector.py:
```diff
@@ -147,7 +148,29 @@ def _show_logging_func():
         self, symbol: str, interval: str, start_datetime: pd.Timestamp, end_datetime: pd.Timestamp
     ) -> pd.DataFrame:
-        @deco_retry(retry_sleep=self.delay)
+        if hasattr(self, 'retry_config'):
```
Will it look simpler  to make retry as a part of the class  interface ( for example, [abstract  attribute](https://stackoverflow.com/questions/2736255/abstract-attributes-in-python) )

igor17400 replied:
@you-n-g but should I define this class interface only for `YahooCollectorBR1d` and `YahooCollectorBR1min` classes?

you-n-g replied:
Maybe we can define

```
class YahooCollector(...):
      retry = 5  # Configuration attribute.  How many times will it try to re-request the data if the network fails.

class YahooCollectorBR(...):
        """"
            The reason to use retry=2 is due to the fact that
            Yahoo Finance unfortunately does not keep track of the majority
            of Brazilian stocks. 
            
            Therefore, the decorator deco_retry with retry argument
            set to 5 will keep trying to get the stock data 5 times, 
            which makes the code to download Brazilians stocks very slow. 
            
            In future, this may change, but for now 
            I suggest to leave retry argument to 1 or 2 in 
            order to improve download speed.
            In order to achieve this code logic an argument called retry_config
            was added into YahooCollectorBR base class
        """
      retry = 2
```

Now the retry is part of the interface, so we don't have to use hasattr to access the retry.

igor17400 replied:
@you-n-g I've added such modifications into commit https://github.com/microsoft/qlib/pull/990/commits/c31380481d189d5b7e05490dd9c662f975dffa97

you-n-g commented:
It looks great now!

SunsetWolf commented:
Hi, @igor17400 
This script is not working right now because of the missing datasource, hopefully the datasource can be updated.
